### ISSUE TITLE
Mypy mistakes instance attrs for class attrs?

### ISSUE THREAD

**Issue by poneill** [python/mypy#15786]
**Bug Report**

Consider the following class definition:

```python
from typing import Any


class Foo:
    @classmethod
    def bar(cls) -> Any:
        return cls.baz  # mypy correctly catches attr-defined error
```
Mypy (1.4.1) correctly notices that `baz` is not defined.  However, if we amend the class definition slightly by adding `baz` as an _instance_ attr like so:

```python
class Foo2:
    def __init__(self, baz: int) -> None:
        self.baz = baz

    @classmethod
    def bar(cls) -> int:
        return cls.baz  # but mypy doesn't catch the attr-defined error here
```

now mypy doesn't spot the error anymore: it seems to be confused about whether `baz` is an instance or a class attr.

**To Reproduce**

Please see playground gist [here](https://mypy-play.net/?mypy=latest&python=3.11&flags=strict&gist=37947d0b493f7dd27816c79918dd0165).

**Expected Behavior**

I would expect mypy to raise an [attr-defined] error for `Foo2` just as for `Foo1`.

**Actual Behavior**
```
main.py:7: error: "type[Foo]" has no attribute "baz"  [attr-defined]
Found 1 error in 1 file (checked 1 source file)
```
[But mypy finds nothing for Foo2.]

**Your Environment**

- Mypy version used: 1.4.1
- Mypy command-line flags:  --strict
- Mypy configuration options from `mypy.ini` (and other config files): NA
- Python version used: 3.11

Thanks!

**Comment by erictraut** [comment]
I think this is a duplicate of #240.

**Comment by AlexWaygood** [comment]
> I think this is a duplicate of #240.

Indeed

**Comment by ikonst** [comment]
Maybe we should split reading and writing? #240 is about writing to the class instance, which @JukkaL proceeded to comment might be a tough call to make.

**Comment by erictraut** [comment]
FWIW, [here](https://microsoft.github.io/pyright/#/mypy-comparison?id=class-and-instance-variable-enforcement) is a description of how I addressed this problem in pyright.

**Comment by poneill** [comment]
> Maybe we should split reading and writing? #240 is about writing to the class instance, which @JukkaL proceeded to comment might be a tough call to make.

I'm far from an expert on this topic, but this issue (is about reading instance attributes as class attributes) does seem distinct from #240 (about writing them).  (This issue is, however, a duplicate of https://github.com/python/mypy/issues/240#issuecomment-1044752777, https://github.com/python/mypy/issues/15307, https://github.com/python/mypy/issues/1097 and others).

The important difference between #240 and this issue, AIUI, is that the snippet in #240 doesn't raise an `AttributeError` at runtime, so mypy is simply following python's counterintuitive behavior in this case.  The snippet in this issue, however, will raise `AttributeError`:

```python
foo2 = Foo2(1)
foo2.baz  # AttributeError: type object 'Foo2' has no attribute 'baz'
```

It sounds like there are deep-seated reasons why this error is hard to catch statically, but Python does indeed consider it an error.
